[lldb] Refactoring JSONTransport into an abstract RPC Message Handler and transport layer.

ashgti opened the pull request:
This abstracts the base Transport handler to have a MessageHandler component and allows us to generalize both JSON-RPC 2.0 for MCP (or an LSP) and DAP format.

This should allow us to create clearly defined clients and servers for protocols, both for testing and for RPC between the lldb instances and an lldb-mcp multiplexer.

This basic model is inspiried by the clangd/Transport.h file and the mlir/lsp-server-support/Transport.h that are both used for LSP servers within the llvm project.

Additionally, this helps with testing by subclassing `Transport` to allow us to simplify sending/receiving messages without needing to use a toJSON/fromJSON and a pair of pipes, see `TestTransport` in DAP/TestBase.h.

ashgti committed:
[lldb] Refactoring JSONTransport into an abstract RPC Message Handler and transport layer.

This abstracts the base Transport handler to have a MessageHandler component and allows us to generalize both JSON-RPC 2.0 for MCP (or an LSP) and DAP format.

This should allow us to create clearly defined clients and servers for protocols, both for testing and for RPC between the lldb instances and an lldb-mcp multiplexer.

This basic model is inspiried by the clangd/Transport.h file and the mlir/lsp-server-support/Transport.h that are both used for LSP servers within the llvm project.

JDevlieghere commented on lldb/include/lldb/Host/JSONTransport.h:
```diff
@@ -54,112 +50,219 @@ class TransportUnhandledContentsError
+/// application, and reading/writing structured messages to it.
+///
+/// Transports have limited thread safety requirements:
+///  - messages will not be sent concurrently
+///  - messages MAY be sent while Run() is reading, or its callback is active
```
```suggestion
///  - Messages will not be sent concurrently.
///  - Messages MAY be sent while Run() is reading, or its callback is active.
```

ashgti replied:
Done.

JDevlieghere commented on lldb/include/lldb/Host/JSONTransport.h:
```diff
@@ -54,112 +50,219 @@ class TransportUnhandledContentsError
+    std::string output = Encode(message);
+    size_t bytes_written = output.size();
+    Status status = m_out->Write(output.data(), bytes_written);
+    if (status.Fail()) {
+      this->Logv("writing failed: {0}", status.AsCString());
+    }
```
Leftover braces from LLDB_LOG I presume.
```suggestion
    if (status.Fail())
      this->Logv("writing failed: {0}", status.AsCString());
```

JDevlieghere commented on lldb/include/lldb/Protocol/MCP/Protocol.h:
```diff
@@ -23,41 +24,59 @@ namespace lldb_protocol::mcp {
 /// A request that expects a response.
 struct Request {
-  uint64_t id = 0;
+  Id id; // Per the spec, this must be a string or number.
```
Maybe move this comment to line 27 so we don't need to repeat it everywhere that used `Id`?

```
/// A Request or Response 'id'. Per the spec, this must be a string or number.
using Id = std::variant<std::string, int64_t>;
```

ashgti replied:
Done, also moved these changes into #153297

JDevlieghere commented on lldb/unittests/TestingSupport/TestUtilities.h:
```diff
@@ -62,6 +62,7 @@ class TestFile {
 
 template <typename T> static llvm::Expected<T> roundtripJSON(const T &input) {
   llvm::json::Value value = toJSON(input);
+  llvm::errs() << "JSON: " << value << "\n";
```
Leftover from debugging?

ashgti replied:
Done.

JDevlieghere reviewed (commented):
Overall this LGTM. I like the design of the Transport class, albeit still very much tied to the protocols it's meant to support. Both DAP and MCP have requests, responses and events, but I don't know how generic that is. Anyway, it should be easily extensible and/or use `std::monostate` if they don't support one of the types. 

The PR includes some unrelated reflowing of comments, might be nice to unstage that. There's some other changes (like the `Id` for MCP) that look like they could go into their own PR, but there might be hidden dependencies.

ashgti commented:
Split out the MCP changes into #153297 Once thats in I'll rebase on that and see if I can revert/simplify anything else.

JDevlieghere commented on lldb/include/lldb/Host/JSONTransport.h:
```diff
@@ -54,112 +50,220 @@ class TransportUnhandledContentsError
+    std::string output = Encode(message);
+    size_t bytes_written = output.size();
+    Status status = m_out->Write(output.data(), bytes_written);
+    if (status.Fail()) {
+      this->Logv("writing failed: s{0}", status.AsCString());
+    }
```
```suggestion
    if (status.Fail())
      this->Logv("writing failed: s{0}", status.AsCString());
```

JDevlieghere commented on lldb/include/lldb/Host/JSONTransport.h:
```diff
@@ -54,112 +50,220 @@ class TransportUnhandledContentsError
+        return;
+      }
+
+      for (const auto &raw_message : *raw_messages) {
```
Nit: type isn't obvious from RHS.
```suggestion
      for (const std::string &raw_message : *raw_messages) {
```

ashgti replied:
Done.

JDevlieghere commented on lldb/include/lldb/Host/JSONTransport.h:
```diff
@@ -54,112 +50,220 @@ class TransportUnhandledContentsError
+          return;
+        }
+
+        if (Evt *evt = std::get_if<Evt>(&*message)) {
+          handler.OnEvent(*evt);
+        } else if (Req *req = std::get_if<Req>(&*message)) {
+          handler.OnRequest(*req);
+        } else if (Resp *resp = std::get_if<Resp>(&*message)) {
+          handler.OnResponse(*resp);
+        } else {
+          llvm_unreachable("unknown message type");
+        }
```
Either no braces or, alternatively you could do something like

```
if (Evt *evt = std::get_if<Evt>(&*message)) {
  handler.OnEvent(*evt);
  continue;
}

if (Req *req = std::get_if<Req>(&*message)) {
  handler.OnRequest(*req);
  continue;
}
```

ashgti replied:
I went with the ifs + `continue`, I think that looks a bit nicer.

JDevlieghere commented on lldb/tools/lldb-dap/tool/lldb-dap.cpp:
```diff
@@ -344,13 +346,14 @@ serveConnection(const Socket::SocketProtocol &protocol, const std::string &name,
         client_failed = true;
-        llvm::errs() << "DAP client " << dap->transport.GetClientName()
-                     << " disconnected failed: "
+        llvm::errs() << "DAP client disconnected failed: "
```
Since you're here:
```suggestion
        WithColor::error() << "DAP client disconnected failed: "
```

ashgti replied:
Done.

JDevlieghere commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -258,36 +259,33 @@ void DAP::SendJSON(const llvm::json::Value &json) {
+  } else if (const Response *resp = std::get_if<Response>(&message)) {
+    // FIXME: After all the requests have migrated from LegacyRequestHandler >
+    // RequestHandler<> this should be handled in RequestHandler<>::operator().
+    if (debugger.InterruptRequested())
+      // If the debugger was interrupted, convert this response into a
+      // 'cancelled' response because we might have a partial result.
+      transport.Response(Response{/*request_seq=*/resp->request_seq,
+                                  /*command=*/resp->command,
+                                  /*success=*/false,
+                                  /*message=*/eResponseMessageCancelled,
+                                  /*body=*/std::nullopt});
+    else
+      transport.Response(*resp);
```
According to the style guide, this does require braces:
```suggestion
    if (debugger.InterruptRequested()) {
      // If the debugger was interrupted, convert this response into a
      // 'cancelled' response because we might have a partial result.
      transport.Response(Response{/*request_seq=*/resp->request_seq,
                                  /*command=*/resp->command,
                                  /*success=*/false,
                                  /*message=*/eResponseMessageCancelled,
                                  /*body=*/std::nullopt});
    } else {
      transport.Response(*resp);
    }
```

JDevlieghere commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -938,88 +933,74 @@ void DAP::ClearCancelRequest(const CancelArguments &args) {
+  llvm::ErrorAsOutParameter ErrAsOutParam(*error);
+  auto cleanup = llvm::make_scope_exit([&]() {
+    // Ensure we're marked as disconnecting when the reader exits.
+    disconnecting = true;
```
I know this was already there, but technically we're racing on `disconnecting` here.

ashgti replied:
Added a lock guard around `disconnecting` and made it private.

JDevlieghere reviewed (approved):
Left some comments with nits but besides that LGTM.

ashgti committed:
Addressing reviewer comments.

ashgti committed:
Refactoring the Transport to not directly run the MainLoop and surface errors to the MessageHandler instead of internally logging and ignoring them.

This allows the MessageHandler to determine what to do when an error occurs.

ashgti commented on lldb/include/lldb/Host/JSONTransport.h:
```diff
@@ -54,112 +50,227 @@ class TransportUnhandledContentsError
+      return status.takeError();
+    }
+
+    status = loop.Run();
```
Updated to have a register function instead of a `Run` function.

ashgti commented:
I refactored this a bit to not have the `Transport` run the `MainLoop`, instead have it register with the loop.

This changed the flow of how errors are handled. Now instead of simply logging any errors I surface them to the `MessageHandler` and it can choose how to handle the error, either log it or whatever.

LMKWYT

labath reviewed (commented):
The main loop changes seem fine. Some other comments below. :)

ashgti committed:
Using gmock for testing message handling.

Also refactored the Transport and MessageHandler classes to have more uniform names, which allows us to use `std::visit` for dispatching to the correct method.

ashgti committed:
Merge remote-tracking branch 'origin/main' into lldb-message-handler

ashgti committed:
Merge remote-tracking branch 'origin/main' into lldb-message-handler

ashgti committed:
Add subsystems initialization to tests.

labath reviewed (approved):
LGTM
